File: src/io/github/bagu/dough/lexer/Lexer.java
package io.github.bagu.dough.lexer;

import io.github.bagu.dough.source.SourceRange;
import io.github.bagu.dough.token.Token;
import io.github.bagu.dough.token.TokenKind;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public final class Lexer {
    private static final char EOF = '\0';
    private static final Map<String, TokenKind> KEYWORDS;

    static {
        KEYWORDS = Map.ofEntries(
                Map.entry("str", TokenKind.STR),
                Map.entry("int", TokenKind.INT),
                Map.entry("float", TokenKind.FLOAT),
                Map.entry("bool", TokenKind.BOOL),
                Map.entry("void", TokenKind.VOID),

                Map.entry("true", TokenKind.TRUE),
                Map.entry("false", TokenKind.FALSE),

                Map.entry("var", TokenKind.VAR),
                Map.entry("func", TokenKind.FUNC),

                Map.entry("if", TokenKind.IF),
                Map.entry("else", TokenKind.ELSE),
                Map.entry("while", TokenKind.WHILE),
                Map.entry("return", TokenKind.RETURN),

                Map.entry("and", TokenKind.AND),
                Map.entry("or", TokenKind.OR),
                Map.entry("not", TokenKind.NOT)
        );
    }

    private final String src;
    private int current = 0;

    public Lexer(String src) {
        this.src = Objects.requireNonNull(src);
    }

    public List<Token> lex() {
        current = 0;
        List<Token> tokens = new ArrayList<>();

        while (!atEnd()) {
            Token token = scanToken();
            if (token.kind() == TokenKind.EOF) break;
            tokens.add(token);
        }

        tokens.add(token(TokenKind.EOF, null, current));

        return tokens;
    }

    private SourceRange range(int start) {
        return new SourceRange(start, current);
    }

    private Token token(TokenKind kind, String lexeme, int start) {
        return new Token(kind, lexeme, range(start));
    }

    private boolean atEnd() {
        return current >= src.length();
    }

    private char advance() {
        return atEnd()
                ? EOF
                : src.charAt(current++);
    }

    private char peek(int offset) {
        int index = current + offset;
        return (index >= src.length())
                ? EOF
                : src.charAt(index);
    }

    private boolean check(char expected) {
        return peek(0) == expected;
    }

    private boolean match(char expected) {
        boolean matches = check(expected);
        if (matches) advance();
        return matches;
    }

    private void skip(int count) {
        for (int i = 0; i < count; i++) {
            advance();
        }
    }

    private Token scanToken() {
        while (true) {
            skipWhitespaces();
            skipComments();

            char peek = peek(0);
            if (Character.isWhitespace(peek) ||
                    (peek == '/' && peek(1) == '*')) {
                continue;
            }
            break;
        }

        if (atEnd()) return token(TokenKind.EOF, null, current);

        int start = current;
        char peek = peek(0);

        if (peek == '"')
            return scanString();

        if (peek == '_' || Character.isLetter(peek))
            return scanIdent();

        if (Character.isDigit(peek))
            return scanNumber();

        skip(1);
        return switch (peek) {
            case '+' -> token(TokenKind.PLUS, "+", start);
            case '-' -> token(TokenKind.MINUS, "-", start);
            case '*' -> token(TokenKind.STAR, "*", start);
            case '/' -> token(TokenKind.SLASH, "/", start);

            case '(' -> token(TokenKind.LPAREN, "(", start);
            case ')' -> token(TokenKind.RPAREN, ")", start);
            case '{' -> token(TokenKind.LBRACE, "{", start);
            case '}' -> token(TokenKind.RBRACE, "}", start);
            case ',' -> token(TokenKind.COMMA, ",", start);
            case ':' -> token(TokenKind.COLON, ":", start);
            case ';' -> token(TokenKind.SEMICOLON, ";", start);

            case '=' -> match('=')
                    ? token(TokenKind.EQ, "==", start)
                    : token(TokenKind.ASSIGN, "=", start);
            case '<' -> match('=')
                    ? token(TokenKind.LE, "<=", start)
                    : token(TokenKind.LT, "<", start);
            case '>' -> match('=')
                    ? token(TokenKind.GE, ">=", start)
                    : token(TokenKind.GT, ">", start);

            case '!' -> {
                if (match('='))
                    yield token(TokenKind.NEQ, "!=", start);
                throw new LexerException("Unexpected character '" + peek + "'", range(start));
            }

            default -> throw new LexerException("Unexpected character '" + peek + "'", range(start));
        };
    }

    private void skipWhitespaces() {
        while (Character.isWhitespace(peek(0))) {
            skip(1);
        }
    }

    private void skipComments() {
        if (check('/') && peek(1) == '*') {
            skip(2);
            int start = current;

            while (!atEnd() && !(check('*') && peek(1) == '/')) {
                skip(1);
            }

            if (atEnd()) throw new LexerException("Unclosed block comment", range(start));
            skip(2);
        }
    }

    private Token scanString() {
        if (!match('"')) throw new AssertionError();

        int start = current;
        StringBuilder sb = new StringBuilder();

        while (!atEnd() && !check('"')) {
            if (check('\\')) {
                char escape = switch (peek(1)) {
                    case '\\' -> '\\';
                    case '"' -> '\"';
                    case 'n' -> '\n';

                    default -> throw new LexerException("Unknown escape", range(current));
                };

                sb.append(escape);
                skip(2);
            } else {
                sb.append(advance());
            }
        }

        if (atEnd()) throw new LexerException("Unclosed string literal", range(start));
        skip(1);

        return token(TokenKind.STR_LIT, sb.toString(), start);
    }

    private Token scanIdent() {
        if (!check('_') && !Character.isLetter(peek(0)))
            throw new AssertionError();

        int start = current;
        StringBuilder sb = new StringBuilder();

        while (!atEnd() && (check('_') || Character.isLetterOrDigit(peek(0)))) {
            sb.append(advance());
        }

        String lexeme = sb.toString();
        TokenKind kind = KEYWORDS.getOrDefault(lexeme, TokenKind.IDENT);
        return token(kind, lexeme, start);
    }

    private Token scanNumber() {
        if (!Character.isDigit(peek(0))) throw new AssertionError();

        int start = current;
        StringBuilder sb = new StringBuilder();
        boolean hasDot = false;

        while (!atEnd() && (check('.') || Character.isDigit(peek(0)))) {
            if (check('.')) {
                if (hasDot)
                    throw new LexerException("Multiple decimal points in float literal", range(start));
                if (!Character.isDigit(peek(1)))
                    throw new LexerException("Expected digit after decimal point", range(start));

                hasDot = true;
            }

            sb.append(advance());
        }

        TokenKind kind = hasDot ? TokenKind.FLOAT_LIT : TokenKind.INT_LIT;
        return token(kind, sb.toString(), start);
    }
}

File: src/lexer.rs
use crate::lexer::token::{Token, TokenKind};
use crate::span::Span;

mod token;

pub(crate) enum Error {
    UnexpectedChar(char, Span),
    UnterminatedString(Span),
    InvalidNumber(String, Span),
}

pub(crate) type Result<T> = std::result::Result<T, Error>;

/// A cursor over source text.
struct Cursor<'a> {
    source: &'a str,
    pos: usize,
    // cache the next char when peeking
    next: Option<char>,
}

impl<'a> Cursor<'a> {
    /// Sentinel to signal the end of the source, used to prevent
    /// `Option<char>` from propagating everywhere.
    const EOF_CHAR: char = '\0';

    fn new(source: &'a str) -> Self {
        Self { source, pos: 0, next: None }
    }

    /// Returns `true` if the cursor is at the end of the source.
    fn is_at_end(&self) -> bool {
        self.pos >= self.source.len()
    }

    /// Returns the next character without advancing.
    ///
    /// # Notes
    /// Returns '\0' if the cursor is at the end of the source.
    fn peek(&mut self) -> char {
        if let Some(c) = self.next {
            return c;
        }

        let c = self.source[self.pos..]
            .chars()
            .next()
            .unwrap_or(Self::EOF_CHAR);

        self.next = Some(c);
        c
    }

    /// Advances the cursor and returns the consumed character.
    ///
    /// # Notes
    /// Returns '\0' if the cursor is at the end of the source.
    fn advance(&mut self) -> char {
        let c = self.peek();

        if c != Self::EOF_CHAR {
            self.pos += c.len_utf8();
        }

        self.next = None;
        c
    }

    /// Advances while `pred` holds and the cursor is not at the end
    /// of the source, returning the consumed slice.
    fn advance_while(&mut self, pred: impl Fn(char) -> bool) -> &str {
        let start = self.pos;

        while !self.is_at_end() && pred(self.peek()) {
            self.advance();
        }

        &self.source[start..self.pos]
    }

    /// Advances and returns `true` if the next character equals `c`.
    fn match_char(&mut self, c: char) -> bool {
        if self.peek() == c {
            self.advance();
            true
        } else {
            false
        }
    }
}

/// A lexer for Dough source text.
pub(crate) struct Lexer<'a> {
    cursor: Cursor<'a>,
    start: usize,
}

impl Lexer<'_> {
    fn new_span(&self) -> Span {
        Span::new(self.start, self.cursor.pos)
    }

    fn new_token(&self, kind: TokenKind) -> Token {
        Token::new(kind, self.new_span())
    }

    /// Advances the cursor and returns `kind`.
    fn single(&mut self, kind: TokenKind) -> TokenKind {
        self.cursor.advance();
        kind
    }

    /// Advances the cursor and returns `two`, advancing the cursor
    /// once again, if the next character equals `c`, otherwise `one`.
    fn double(&mut self, c: char, one: TokenKind, two: TokenKind) -> TokenKind {
        self.cursor.advance();
        if self.cursor.match_char(c) {
            two
        } else {
            one
        }
    }
}

impl<'a> Lexer<'a> {
    /// Tokenizes `source`, returning a list of tokens or [`Error`] on failure.
    pub(crate) fn tokenize(source: &'a str) -> Result<Vec<Token>> {
        Self::new(source).tokenize_inner()
    }

    fn new(source: &'a str) -> Self {
        Self { cursor: Cursor::new(source), start: 0 }
    }

    fn tokenize_inner(mut self) -> Result<Vec<Token>> {
        let mut tokens = Vec::new();

        while !self.cursor.is_at_end() {
            self.cursor.advance_while(|c| c.is_whitespace());
            self.start = self.cursor.pos;

            if self.cursor.is_at_end() {
                break;
            }

            tokens.push(self.scan_token()?);
        }

        tokens.push(self.new_token(TokenKind::Eof));
        Ok(tokens)
    }

    /// Scans the next token from the current position.
    ///
    /// # Panics
    /// If the cursor is at the end of the source.
    fn scan_token(&mut self) -> Result<Token> {
        assert!(!self.cursor.is_at_end());

        let kind = match self.cursor.peek() {
            // punctuation
            '(' => self.single(TokenKind::LParen),
            ')' => self.single(TokenKind::RParen),
            '{' => self.single(TokenKind::LBrace),
            '}' => self.single(TokenKind::RBrace),
            '[' => self.single(TokenKind::LBrack),
            ']' => self.single(TokenKind::RBrack),
            ':' => self.single(TokenKind::Colon),
            ';' => self.single(TokenKind::Semicolon),
            ',' => self.single(TokenKind::Comma),
            '.' => self.single(TokenKind::Dot),

            // 1 char operators
            '+' => self.single(TokenKind::Plus),
            '-' => self.single(TokenKind::Minus),
            '*' => self.single(TokenKind::Star),
            '/' => self.single(TokenKind::Slash),

            // 1-2 char operators
            '=' => self.double('=', TokenKind::Eq, TokenKind::EqEq),
            '<' => self.double('=', TokenKind::Lt, TokenKind::Le),
            '>' => self.double('=', TokenKind::Gt, TokenKind::Ge),
            '!' => {
                // essentially the same as self.double but `!` does not exist
                self.cursor.advance();
                if self.cursor.match_char('=') {
                    TokenKind::Ne
                } else {
                    return Err(Error::UnexpectedChar('!', self.new_span()))
                }
            },

            // other
            c if c.is_ascii_digit() => self.scan_number()?,
            // only allow alphabet/_ on 1st ident char
            c if c.is_alphabetic() || c == '_' => self.scan_ident()?,
            '"' => self.scan_str()?,

            c => return Err(Error::UnexpectedChar(c, self.new_span()))
        };

        Ok(self.new_token(kind))
    }

    /// Scans a number literal.
    ///
    /// # Panics
    /// If the current character is not an ascii digit.
    fn scan_number(&mut self) -> Result<TokenKind> {
        assert!(self.cursor.peek().is_ascii_digit());

        let s = self.cursor.advance_while(
            |c| c.is_ascii_digit() || c == '.')
            .to_string();

        // validate after for simplicity as it is a micro-optimization
        let dot_count = s.chars()
            .filter(|&c| c == '.')
            .count();

        let span = self.new_span();

        match dot_count {
            0 => s.parse::<i64>()
                .map(TokenKind::Int)
                .map_err(|_| Error::InvalidNumber(s, span)),

            1 => s.parse::<f64>()
                .map(TokenKind::Float)
                .map_err(|_| Error::InvalidNumber(s, span)),

            _ => Err(Error::InvalidNumber(s, span))
        }
    }

    /// Scans an identifier or keyword.
    ///
    /// # Panics
    /// If the current character is not `_` or alphanumeric.
    fn scan_ident(&mut self) -> Result<TokenKind> {
        let c = self.cursor.peek();
        assert!(c.is_alphanumeric() || c == '_');

        let s = self.cursor.advance_while(
            |c| c.is_alphanumeric() || c == '_')
            .to_string();

        let kind = match s.as_str() {
            "true" => TokenKind::Bool(true),
            "false" => TokenKind::Bool(false),
            "var" => TokenKind::Var,
            "func" => TokenKind::Func,
            "return" => TokenKind::Return,
            "if" => TokenKind::If,
            "else" => TokenKind::Else,
            "while" => TokenKind::While,
            "not" => TokenKind::Not,
            "and" => TokenKind::And,
            "or" => TokenKind::Or,
            _ => TokenKind::Ident(s)
        };

        Ok(kind)
    }

    /// Scans a string literal.
    ///
    /// # Panics
    /// If the current character is not `"`.
    fn scan_str(&mut self) -> Result<TokenKind> {
        assert_eq!(self.cursor.advance(), '"');

        let s = self.cursor.advance_while(
            |c| c != '"').to_string();

        if self.cursor.match_char('"') {
            Ok(TokenKind::Str(s))
        } else {
            Err(Error::UnterminatedString(self.new_span()))
        }
    }
}
